README
Validator version 0.01
======================

form_validator.xsd - in t/xmlData

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2000 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Validator.

0.01  Sat Jan  1 00:41:50 2000
	- original version; created by h2xs 1.23 with options
		-X -n Validator



Makefile.PL
use 5.008007;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Validator',
    VERSION_FROM      => 'lib/Validator.pm', # finds $VERSION
    PREREQ_PM	  => { 
    	'JSON::XS' => 0,
    	'Class::Accessor::Fast' => 0,
    	'Mail::RFC822::Address' => 0,
    	'AGAVA::AGE::Framework::Library::XML::XPath::Cached' => 0,
    	'XML::XPath' => 0,
    	'Cache::FileCache' => 0,
    	'XML::XPath::XMLParser' => 0,
    	'XML::Validator::Schema' => 0,
    	'XML::Simple' => 0,
    	'XML::SAX::ParserFactory' => 0
    },	
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Validator.pm', # retrieve abstract from module
       AUTHOR         => 'Alex Nosoff <plcgi1@gmail.com>') : ()),
);


lib/Validator/Rules/Base.pm
package Validator::Rules::Base;

use 5.008007;
use strict;
use warnings;

use  Mail::RFC822::Address qw(valid);

our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

sub new {
	my $class = shift;
	my $this = {};
	bless $this,$class;
	return $this;
}

sub integer {
	my $this = shift;
	my $fieldValue = shift;
	my $res = ( $fieldValue =~/^\d+$/ );
	return $res;		
}
	
sub pattern	{
	my $this = shift;
	my $fieldValue 	= shift;
	my $pattern		= shift;
	
	my $res = ( $fieldValue =~ /$pattern/ );
	return $res;		
}
		
sub ip {
	my $this = shift;
	my $fieldValue 	= shift;
	
	my $res = ( $fieldValue=~/^(\d|[01]?\d\d|2[0-4]\d|25[0-5])\.(\d|[01]?\d\d|2[0-4]\d|25[0-5])\.(\d|[01]?\d\d|2[0-4]\d|25[0-5])\.(\d|[01]?\d\d|2[0-4]\d|25[0-5])$/ );
	return $res;
}	

sub email {
	my $this = shift;
	my $fieldValue 	= shift;
	
	my $res = valid($fieldValue);
	
	return $res;
}	

sub maxlength {
	my $this = shift;
	my $fieldValue 	= shift;
	my $maxLength 	= shift;
	
	unless ( $fieldValue =~/^[\p{IsDigit}]+$/) {
		$fieldValue = length($fieldValue);
	}
		
	my $res = ( $fieldValue <= $maxLength ) ;
	return $res;		
}
		
sub minlength {
	my $this = shift;
	my $fieldValue 	= shift;
	my $minLength 	= shift;
	
	unless ( $fieldValue =~/^[\p{IsDigit}]+$/) {
		$fieldValue = length($fieldValue);
	}
		
	my $res = ( $fieldValue >= $minLength ) ;
	return $res;
}

sub anyText {
	my $this = shift;
	my $fieldValue 	= shift;
	
	#my $re = q/^[\p{IsAlpha}|\p{IsDigit}|\p{IsSpace}|\p{IsS}|\p{IsP}|\p{IsZ}|\p{InCyrillic}]+$/;
	
	my $re = '.*';
	
	my $res = $this->pattern($fieldValue,$re);
	
	return $res;
}

sub equals {
	my $this 	= shift;
	my $value1 	= shift;
	my $value2 	= shift;
	
	my $res = $value1 eq $value2;
	
	return $res;
}

sub notEquals {
	my $this 	= shift;
	my $value1 	= shift;
	my $value2 	= shift;
	
	my $res = $this->equals($value1,$value2);
	
	return !$res;
}
1;
__END__

=head1 NAME

Validator::Rules::Base - Perl extension for field values validation

=head1 SYNOPSIS

  use Validator::Rules::Base;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Validator::Rules::Base, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.


=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

A. U. Thor, E<lt>a.u.thor@a.galaxy.far.far.awayE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2000 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/AGAVA/AGE/Framework/Library/XML/XPath/Cached.pm
package AGAVA::AGE::Framework::Library::XML::XPath::Cached;
use strict;
use warnings;
use base "XML::XPath";

##############################################################################
# whatever in Cache:xx hierarchy
our $CACHE_MODULE = "Cache::FileCache";
use Cache::FileCache;

# namespace equals to package name with '::' replace with '/', so directories
# are created
(our $CACHE_NAMESPACE = __PACKAGE__) =~ s|::|/|xg;
##############################################################################

use File::stat;
use XML::XPath::XMLParser;
use XML::Validator::Schema;
use XML::SAX::ParserFactory;

use Storable qw (freeze);
use XML::Simple;

use Carp;

=pod

=head1 NAME

AGAVA::AGE::Framework::Library::XML::XPath::Cached - XPath sublass to cache
parsed XML content.

=head1 SYNOPSIS

	$xpath = new AGAVA::AGE::Framework::Library::XML::XPath::Cached (
					filename => "file.xml", xsdFilename => "file.xsd");
					
=head1 DESCRIPTION

This class is a drop-in replacement for XML::XPath designed to:

=over

=item

speed things up by caching parsed XML content

=item

validate XML content against XML schema

=back


=head1 METHODS

=over 

=cut



=pod

=item B<new (filename =E<gt> file.xml, xsdFilename =E<gt> file.xsd)>

Constructor. Options are:

=over

=item B<filename> - mandatory

input XML file (if '/' is not present in its name, it's not unique
cache-wide and an exception is thrown (Cwd::realpath() is sometimes buggy and
returns undef)

=item B<xsdFilename> - optional

XML schema file to validate XML content against

=item B<xmlSimple> -  hashref

what to pass to XML::Simple::XMLin() - ForceArray=1 and KeyAttr=>[] are
enforced to guarantee XML is treated as safely as possible

=back

=cut

sub new (%)
{
    my ($class, %args) = @_;
    $class = ref ($class) || $class;
    
    # so far, only filename is accepted (no ioref etc.)
    if (!defined ($args{filename}) || ($args{filename} eq ""))
    {
    	croak ("Please pass non-empty filename=>file.xml argument");
    }
	
    
    # absolute path provides cache key uniqueness if filename is passed
    # as e.g. '../config.xml' (from different current dirs)
    # TODO: why does Cwd::realpath() return undef??
    my $xmlFile = $args{filename};
    
    if ($xmlFile !~ /^ \/ /x)
    {
    	croak ("Filename should start with '/': '$xmlFile'");
    }
    
#    my $xmlFile = realpath ($args{filename});
#    if (!defined ($xmlFile))
#    {
#    	croak ("Can't build realpath for file '$xmlFile'");
#    }
    
    # extension point for other types, such as xml://
    # (if $args{xml} is present)
    my $rootPrefix = "file://$xmlFile";
	
	my $cache = $CACHE_MODULE->new ({ namespace => $CACHE_NAMESPACE });
	my $rootNodeKey 	 = "$rootPrefix/xpathRootNode";
	my $xmlFileSizeKey 	 = "$rootPrefix/xmlFileSize";
	my $xmlFileMtimeKey  = "$rootPrefix/xmlFileMtime";
	my $xmlSimpleKey 	 = "$rootPrefix/xmlSimple";
	
	# true - values in cache are relevant and should be used
	my $isCacheValid = 1;
	
	# current file size the same as the one used to store parsed data?
	my $fileSize = $cache->get ($xmlFileSizeKey);
	
	# current file mtime the same as the one used to store parsed data?
	my $fileMtime = $cache->get ($xmlFileMtimeKey);
	
	my $fileStat = stat ($xmlFile);
	
	# no record in cache or values differ
	if (!defined ($fileStat)
		||
		!defined ($fileSize) || ($fileSize != $fileStat->size())
		||
		!defined ($fileMtime) || ($fileMtime != $fileStat->mtime()) )
	{
		$isCacheValid = 0;
	}
	
	
	my $rootNode = $cache->get ($rootNodeKey);
	my $xmlSimple = $cache->get ($xmlSimpleKey);
	
	# nothing or incomplete info in cache (i.e. if we add something to
	# cache parameter set, cache upgrade will be seamless, as cache will
	# become invalid)
	if (!defined ($rootNode) || !defined ($xmlSimple))
	{
		$isCacheValid = 0;
	}
	
	# populate variables and store them in cache
	if ($isCacheValid)
	{
#		die ("Reading from cache");
	}
	
	else
	{
		# validate XML file against XML schema
	    if (defined ($args{xsdFilename}))
	    {
			# XML::Validator::Schema accepts only file names not file content
			my $handler = new XML::Validator::Schema (
										file  => $args{xsdFilename},
										cache => 1);

			my $parser = XML::SAX::ParserFactory->parser (Handler => $handler);
			
		    # dies on error
			eval { $parser->parse_file ($xmlFile) };
		
		    if ($@ ne "")
		    {
				croak ("Invalid syntax: $xmlFile: $@");
		    }
	    }
		    
	    # parse XML file and store result (xpath nodeset) in cache
		my $parser = new XML::XPath::XMLParser (filename => $xmlFile);
		$rootNode = $parser->parse();
		
		$cache->set ($xmlFileSizeKey,  $fileStat->size());
		$cache->set ($xmlFileMtimeKey, $fileStat->mtime());
		$cache->set ($rootNodeKey, $rootNode);
		
		#
		# store XML converted into hash - for toXMLSimple();
		# for the sake of compatibility, ForceArray=1 (i.e. all tags become
		# arrays )and KeyAttr=[] (i.e. no array folding - array-to-hash
		# conversion - based on name/key/id fields)
		#
		$xmlSimple = XMLin ($xmlFile, ForceArray => 1, KeyAttr => []);
		$cache->set ($xmlSimpleKey, $xmlSimple);
	}

	
	my $self = $class->SUPER::new (context => $rootNode);
	
	# cache always contains the parsed XML nodeset after new(), as relevant
	# data is stored there if it was not there before
    $self->{caching}{cache} = $cache;
    
    # it's dangerous to create own class fields, as they can overlap with
    # the parent's fields, but we have to store some data to access cache
    # in find() - at least we use a certain common prefix for own hash
    $self->{caching}{rootPrefix} = $rootPrefix;
    
    # parsed result
    $self->{caching}{xmlSimple} = $xmlSimple;
    
	return $self;
}



=pod

=item B<toXMLSimple()>

Pipes XML file to XML::Simple, returns what XMLin returns.

NOTE: ForceArray=1 is always passed, so all tags are arrayrefs!
Also, KeyAttr=[] is passed to disable array folding based on attr names.

=cut

sub toXMLSimple()
{
	my ($self) = @_;
	return $self->{caching}{xmlSimple};
}



=pod

=item B<find>

Method borrowed from parent class and patched to use cache, if any.
See parent class for method description.

NOTE: for some reason, in real life projects nodeset caching turned out to be
SLOWER, than no caching, so currently this method DOESN'T USE CACHING.

=cut

# rename info find() to override parent method
# (currently Whois stops working if it's done)
sub findCached {
    my $self = shift;
    my $path = shift;
    my $context = shift;
    die "No path to find" unless $path;
    
    if (!defined $context) {
        $context = $self->get_context;
    }
    if (!defined $context) {
        # Still no context? Need to parse...
        my $parser = XML::XPath::XMLParser->new(
                filename => $self->get_filename,
                xml => $self->get_xml,
                ioref => $self->get_ioref,
                parser => $self->get_parser,
                );
        $context = $parser->parse;
        $self->set_context($context);
#        warn "CONTEXT:\n", Data::Dumper->Dumpxs([$context], ['context']);
    }
    
	# PATCH START
    my $parsed_path = $self->{path_parser}->parse($path);
	# PATCH END
	
#    warn "\n\nPATH: ", $parsed_path->as_string, "\n\n";

     
#    warn "evaluating path\n";
	# PATCH START
	my $nodeset;
	
	# cache always exists - see new() - but we check it for safety
	if (!defined ($self->{caching}{cache}))
	{
		die ("No cache found - did you call new() ?");
	}
		
	# it's better to use the $parsed_path->as_string() (not $path), as it's
	# canonical, but path conversion takes time
	my $key = $self->{caching}{rootPrefix}
			. "/"
			. $path;
			
	$nodeset = $self->{caching}{cache}->get ($key);
	
	# nodeset not found in cache - let parent execute the search query
	if (!defined ($nodeset))
	{
	    $nodeset = $self->SUPER::find ($path, $context);
	    $self->{caching}{cache}->set ($key, $nodeset);
#	    warn "Storing in cache: key $key, no. of elements: " . $nodeset->size();
	}
	
	else
	{
#	    warn "Got from cache: key $key, no. of elements: " . $nodeset->size();
	}
	
	return $nodeset;
#    return $parsed_path->evaluate($context);
	# PATCH END
}

#close method list

=pod

=back

=cut

1;


lib/Validator.pm
package Validator;

use 5.008007;
use strict;
use warnings;

use JSON::XS qw(to_json);
use AGAVA::AGE::Framework::Library::XML::XPath::Cached;

our %EXPORT_TAGS = (
	'all' => [
		qw(

		  )
	]
);

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);

our $VERSION = '0.01';

use Validator::ErrorCode;
use Validator::Rules::Base;
use Class::Accessor::Fast;
use base 'Class::Accessor::Fast';
__PACKAGE__->mk_accessors(qw/fields errorCount errorCode/);

sub new {
	my $class = shift;
	my %opt = @_;
	
	my $this = $class->SUPER::new();
	
	bless $this,$class;
	
	return $this;
}

sub clear {
	my $this = shift;
	
	$this->fields([]);
	$this->{errFields} = undef;	
}

sub isValid {
	my $this = shift;

	my $rulesObj = Validator::Rules::Base->new();

	foreach my $f ( @{ $this->fields } ) {
		my $fieldName  = $f->{name};
		my $required   = $f->{required};
		my $fieldValue = $f->{value};
		my $rules      = $f->{rules};
		my $count      = 0;

		# checking on required
		if ( $required && $required == 1 && length($fieldValue) < 1 ) {
			$f->{error} = "Param $fieldName required" unless $f->{error};
			$this->appendErrField($f);
			next;
		}    # END if ( $required eq '1' && !length($fieldValue) )
		foreach my $r (@$rules) {
			my $func = $r->{rule};
			
			next if ( !$required && !$fieldValue );
			
			my $res = $rulesObj->$func( $fieldValue, $r->{param} );
			if ( !$res ) {
				$f->{error} = "Wrong format for $fieldName" unless $f->{error};
				$this->appendErrField($f);
			}
		}    # END foreach my $r ( keys %$rules )
		
		if ( !$this->{errFields} ) {
			$f->{value} = _filter($f->{value});
		}
	}    # END foreach my $f ( @{$this->fields} )
	my $errors = ++$#{ $this->{errFields} };
	if ( $errors > 0 ) {
		my $err = Validator::ErrorCode->new();

		$err->errorCount($errors);
		$err->errorFields( $this->{errFields} );
		$err->errorMsg();

		return $err;
	}    # END if( $this->errCount() > 0 )

	return 1;
}

sub xmlCached {
	my $this = shift;
	my %opt  = @_; # ( xmlFile => 'path/to/xml/file', xsdFile => '/path/to/xsd/file', values => { fieldName => fieldValue } )
	
	foreach ( qw( xmlFile xsdFile ) ) {
		die "Param $_ is required" unless $opt{$_};
	} 
	
	# hopefully not having to parse the file speeds things up remarkably
	my $xmlCached = AGAVA::AGE::Framework::Library::XML::XPath::Cached->new(
		filename    => $opt{xmlFile},    
		xsdFilename => $opt{xsdFile}
	);

	my $config = $xmlCached->toXMLSimple();
	
	$opt{convertMethod} = \&convertXMLSimpleToValidator unless $opt{convertMethod};
	
	$config = &{ $opt{convertMethod} }( $config );
	
	return $config;
}

sub appendField {
	my $this  = shift;
	my $field = shift;

	push @{ $this->{fields} }, $field;
}

sub appendErrField {
	my $this  = shift;
	my $field = shift;

	push @{ $this->{errFields} }, $field;
}

sub funcIsValidAsJS {
	my $this     = shift;
	my $formName = shift || '' ;

	return unless $this->fields();

	my $fieldsAsJSON = $this->fieldsAsJSON();

	my $funcName = $formName . '_JSValidator';
	my $func     = qq~
<script language="javascript">
    window.$funcName = function() {
	    var validator = new Validator;
	    var fields = $fieldsAsJSON;
	    validator.SetForm(fields, '$formName');
		validator.Process();
		return validator.success;
    }
</script>
~;

	return $func;
}

sub fieldsAsJSON {
	my $this = shift;

	return unless $this->fields();

	my $json = to_json( $this->fields() );

	return $json;
}

sub convertXMLSimpleToValidator {
	my $param = shift;
	
	my @res;
	foreach my $field ( @{$param->{Field}} ) {
		my (@rules,$hashref);
		
		foreach my $rule ( @{$field->{Rules}->[0]->{Rule}} ) {
			$hashref = {
				rule => $rule->{name},
			};
			$hashref->{param}	= $rule->{Param}->[0] if $rule->{Param}; 
			push @rules, $hashref;	
		}
		$hashref = {
			name	=>	$field->{Name}->[0],
			value	=>	$field->{Value}->[0],
			error	=>	$field->{ErrorString}->[0],
			rules	=>	\@rules
		};
		
		if ( $field->{Required}->[0] eq 'true') {
			$hashref->{required} = 1;	
		} 
		else {
			$hashref->{required} = 0;
		}
		
		push @res,$hashref;
	}
	
	return \@res;
}

#------------------------------------------------------------------------------
sub _filter {
    my $str = shift;

    if ( ref ( $str ) eq 'ARRAY' ) {
        foreach ( @{$str} ) {
            my $regexp = qr/\0|#|\&|"|<|>|\(|\)|\|/;
            $str =~ s/($regexp)/_translateSymbols($1)/ge;
        }
    }
    else {
        my $regexp = qr/\0|#|\&|"|<|>|\(|\)|\|/;
        $str =~ s/($regexp)/_translateSymbols($1)/ge if $str;
    }

    return $str;
}

#------------------------------------------------------------------------------
sub _translateSymbols {
    my $sym = shift;
    if ( $sym eq "\0" ) { $sym= ''; }
    elsif ( $sym eq '#' ) { $sym = '&#35;'; }
    elsif ( $sym eq '&' ) { $sym = '&#38;'; }
    elsif ( $sym eq '"' ) { $sym = '&quot;'; }
    elsif ( $sym eq '<' ) { $sym = '&lt;'; }
    elsif ( $sym eq '>' ) { $sym = '&gt;'; }
    elsif ( $sym eq '|' ) { $sym = '&brvbar;'; }
    elsif ( $sym eq '(' ) { $sym = '&#40;'; }
    elsif ( $sym eq ')' ) { $sym = '&#41;'; }

    return $sym;
}
#------------------------------------------------------------------------------
1;

1;
__END__

=head1 NAME

Validator - Input params validator

=head1 SYNOPSIS

  	use Validator;
  	
  	my $fields = [
		{
			name		=>	'Integer',
			error		=>	'Bad format for Integer',
			value		=>	43,
			rules	=>	[
				{ rule => 'integer' },
				{ rule => 'maxlength', param => 1 },
			]
		},
		{ ... }
	];
	
	my $validator = Validator->new();
	$validator->fields($fields);
	my $valid = $validator->isValid();
	
	if ( ref $valid eq 'Validator::ErrorCode' ) {
		# error handling
		$valid->errorCode();
		# or 
		$valid->errorMsg();
	}		

in JSON

# Example of array for validator settings
#	fields =
#	[
#		{
#			name: 'child_frm_1_txt1',
#			required: 1,
#			error: ErrorMessage,
#			value: value
#			rules: [
#				{ rule: 'integer' },
#				{ rule: 'maxlength', param: 3 }
#			]
#		},
#		{
#			name: 'child_frm_1_txt2',
#			required: 0,
#			rules: [
#				{ rule: 'email' },
#			]
#		},
#		{
#			name: 'child_frm_2_txt1',
#			required: 1,
#			rules: [
#				{ rule: 'datetime', param:  'YYYY-MM-DD hh:mm'  }
#			]
#		},
#		{
#			name: 'child_frm_2_txt2',
#			required: 1,
#			rules: [
#				{ rule: 'minlength', param: 2 },
#				{ rule: 'maxlength', param: 5 }
#			]
#		}
#	]
#*/

=head1 DESCRIPTION

Class for input method validation by rules from Validator::Rules::Base 

=head2 EXPORT

TODO

=head1 SEE ALSO

TODO

=head1 AUTHOR

Alex Nosoff E<lt>plcgi1@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2000 by Alex Nosoff

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.


=cut
